core/io/json.cpp

```diff
@@ -422,23 +423,31 @@ Error JSON::_parse_object(Dictionary &object, const CharType *p_str, int &index,
 			}
 
 			if (token.type != TK_STRING) {
-
 				r_err_str = "Expected key";
 				return ERR_PARSE_ERROR;
 			}
 
-			key = token.value;
+			// The object key may contain a serialized variant, parse it
+			VariantParser::StreamString ss;
+			ss.s = token.value;
+
+			err = VariantParser::parse(&ss, key, r_err_str, line);
```

Comment:
`token.value` holds our previously serialized var as string, so we can parse it now, the value is stored in `key` which will be inserted on next token.